### ISSUE TITLE
Prune excessive memory consumption

### ISSUE THREAD

**Issue by pdf** [gilbertchen/duplicacy#146]
I have a 1GB RAM machine that generates one revision per day, with a total local storage volume of 12GB, and a retention policy of `7:30`, and `0:120`.  Attempting to perform a prune operation with ~900MB free RAM results in the process being OOM killed after it consumes all available memory.

**Comment by gilbertchen** [comment]
Are you using the default chunk size of 4MB?  The most memory consuming part by the prune command is just a bunch of maps to store unreferenced chunks (one per revision), and the maps shouldn't be too large with the default size.

I wonder if you can run some memory profiling using the [pprof](https://golang.org/pkg/net/http/pprof/) package.  Here is the output from my repository that has way more revisions (1 reversion every hour, retention policy is 0:18000, 7:30, 1:3, and the storage space is about 10GB):

```
~/zincbox/go/src/github.com/gilbertchen/duplicacy$go tool pprof http://localhost:6060/debug/pprof/heap
Fetching profile from http://localhost:6060/debug/pprof/heap
(pprof)  top
65738.51kB of 65738.51kB total (  100%)
Dropped 51 nodes (cum <= 328.69kB)
Showing top 10 nodes out of 24 (cum >= 16384kB)
      flat  flat%   sum%        cum   cum%
   32768kB 49.85% 49.85%    32768kB 49.85%  bytes.makeSlice
19456.59kB 29.60% 79.44% 19456.59kB 29.60%  runtime.rawstringtmp
13001.85kB 19.78% 99.22% 32458.45kB 49.38%  github.com/gilbertchen/duplicacy/src.(*Snapshot).LoadChunks
  512.06kB  0.78%   100%   512.06kB  0.78%  crypto/sha256.New
         0     0%   100%    32768kB 49.85%  bytes.(*Buffer).Grow
         0     0%   100%    32768kB 49.85%  bytes.(*Buffer).grow
         0     0%   100%   512.06kB  0.78%  crypto/hmac.New
         0     0%   100% 49354.51kB 75.08%  github.com/gilbertchen/cli.(*App).Run
         0     0%   100% 49354.51kB 75.08%  github.com/gilbertchen/cli.Command.Run
         0     0%   100%    16384kB 24.92%  github.com/gilbertchen/duplicacy/src.(*ChunkDownloader).Download
```

**Comment by pdf** [comment]
I am using the default chunk size, however is there a minimum of one chunk per file (so, many small files would blow out the number of chunks)?

I pruned that snapshot from another machine to confirm that it would work correctly, so I may need to accumulate some new revisions to test this again, but I'll provide pprof output when I have it available.

I had a cursory look at the code, so my understanding may not be correct and I apologise if I've got this wrong, but it looks like currently it builds a map of _all_ chunks in the _repository_.  Seems like this would only be necessary if _exhaustive_ was specified, otherwise, could this not work the other way - by building a map of the chunks from the deletion candidates, then checking whether any exist in other snapshots (so only the candidate chunks are held in memory)?

**Comment by gilbertchen** [comment]
That could save some memory, but building a map of all chunks is more efficient, because otherwise you would need to check for every snapshot to keep if any of those candidate chunks exist in that snapshot.  The map for all chunks should not take a lot of space, since it is basically a map of 64-byte strings.

**Comment by pdf** [comment]
I'll have to try and get some profiling data to see where all the memory goes then.

But, I don't follow your efficiency statement.  Why is it more efficient to store all chunks in memory, and check the candidate chunks against those, versus storing the candidate chunks (generally the smaller set), and walking the rest without storing them, just doing a map key lookup on the candidate map?  Should be the same time complexity, but better memory efficiency, no?  Obviously trying to compare the other way, checking each candidate against all chunks would be a terrible idea - that would make the comparison O(n^m) instead of O(n).

**Comment by gilbertchen** [comment]
I misunderstood your approach.  Yes, it would be more space efficient to check every referenced chunk for every snapshot to be kept agains the set of chunks to be deleted, but it would be less speed efficient since one chunk may be referenced by many snapshots and by combining all referenced chunks into one big map it is guaranteed to check each chunk only once.

**Comment by pdf** [comment]
I don't believe there's any difference in performance, since the check is a map lookup, and you need to perform the same work currently, to build the referenced map.

**Comment by gilbertchen** [comment]
I think you're right.  The is certainly an optimization that we ca do.

**Comment by pdf** [comment]
The last `pprof` I managed to capture before all memory was consumed and the OOM killer kicked in:

```
Showing nodes accounting for 496.11MB, 99.90% of 496.61MB total
Dropped 4 nodes (cum <= 2.48MB)
Showing top 10 nodes out of 22
      flat  flat%   sum%        cum   cum%
  373.20MB 75.15% 75.15%   373.20MB 75.15%  github.com/gilbertchen/duplicacy/src.(*Snapshot).LoadChunks github.com/gilbertchen/duplicacy/src/duplicacy_snapshot.go
   51.02MB 10.27% 85.42%    51.02MB 10.27%  github.com/minio/blake2b-simd.New github.com/minio/blake2b-simd/blake2b.go
   33.50MB  6.75% 92.17%    33.50MB  6.75%  bytes.makeSlice /usr/lib/go/src/bytes/buffer.go
      15MB  3.02% 95.19%       15MB  3.02%  encoding/hex.EncodeToString /usr/lib/go/src/encoding/hex/hex.go
   11.35MB  2.28% 97.47%   460.07MB 92.64%  github.com/gilbertchen/duplicacy/src.(*SnapshotManager).GetSnapshotChunks github.com/gilbertchen/duplicacy/src/duplicacy_snapshotmanager.go
       5MB  1.01% 98.48%        5MB  1.01%  github.com/minio/blake2b-simd.(*digest).Sum github.com/minio/blake2b-simd/blake2b.go
    4.50MB  0.91% 99.39%    75.52MB 15.21%  github.com/gilbertchen/duplicacy/src.(*Config).GetChunkIDFromHash github.com/gilbertchen/duplicacy/src/duplicacy_config.go
    2.54MB  0.51% 99.90%   463.11MB 93.25%  github.com/gilbertchen/duplicacy/src.(*SnapshotManager).PruneSnapshots github.com/gilbertchen/duplicacy/src/duplicacy_snapshotmanager.go
         0     0% 99.90%    32.01MB  6.45%  bytes.(*Buffer).Grow /usr/lib/go/src/bytes/buffer.go
         0     0% 99.90%    33.50MB  6.75%  bytes.(*Buffer).grow /usr/lib/go/src/bytes/buffer.go
```

Looks like the loop here:
https://github.com/gilbertchen/duplicacy/blob/6bffef36bf0b5295170d37f35ff548916e8955c9/src/duplicacy_snapshotmanager.go#L1854
which ends up in here, storing all the hashes for each snapshot on the snapshot object:
https://github.com/gilbertchen/duplicacy/blob/6bffef36bf0b5295170d37f35ff548916e8955c9/src/duplicacy_snapshot.go#L351
means that the outer loop requires storing every chunk hash in the repository in memory, for the snapshots that aren't to be deleted, when all we want is a single copy of the chunk IDs.

I think that even if we tried to implement my suggestion from earlier in the thread, there would need to be internal changes so that `GetSnapshotChunks` -> `LoadChunks` doesn't result in all the hashes being stored permanently on `snapshot.ChunkHashes`, otherwise high memory consumption will occur any time the snapshots are iterated, and their chunks retrieved.

**Comment by gilbertchen** [comment]
The main problem here is the each snapshot has their ChunkHashes fully populated when constructing the map of all chunks.  I think snapshot.ChunkHashes can be cleared right after all the chunk hashes from one snapshot have been put into the map.  So if you insert this line between lines 1865 and 1866 (right before existing the loop at 1854):
```
snapshot.ChunkHashes = nil
```
This should release the memory used by the chunk hash list.

Would you like to give it a try?

**Comment by pdf** [comment]
> The main problem here is the each snapshot has their ChunkHashes fully populated when constructing the map of all chunks

Yes, that's what I was trying to explain.

> `snapshot.ChunkHashes = nil`

This is a bit of a hack, but it mostly works around the problem, in that it allowed this client to succeed in a prune operation, however memory consumption still hit around 400MB.  I suspect that allocation of the memory for the chunk hashes in each snapshot could be avoided by introducing a non-caching alternative (or argument) for `GetSnapshotChunks`/`LoadChunks` that doesn't store the chunk hashes at all, but instead just encodes and returns the chunk IDs on the fly.  Combined with the earlier suggestion of reversing the comparison direction, memory usage should be reduced dramatically.

**Comment by s4y** [comment]
I'm running into super high memory consumption on prune, too, and `duplicacy_snapshotmanager.go` has changed enough that I'm not 100% comfortable hacking in the suggested workaround.

Here's a prune that's had a few minutes to warm up using 16GB RAM (it gets OOM killed if I leave it running):

```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      7304  111 51.5 20897500 16931240 pts/3 Sl+ 09:05  15:22 /home/s4y/duplicacygo/bin/duplicacy -profile 127.0.0.1:65123 prune -d -keep 0:360 -keep 30:180 -keep 7:30 -keep 1:7
```

Here's the pprof output (which only accounts for about half the memory — normal?):

```
Showing nodes accounting for 7983.55MB, 99.53% of 8021.58MB total
Dropped 48 nodes (cum <= 40.11MB)
Showing top 10 nodes out of 11
      flat  flat%   sum%        cum   cum%
    7545MB 94.06% 94.06%     7545MB 94.06%  github.com/gilbertchen/duplicacy/src.(*Snapshot).LoadChunks /home/s4y/duplicacygo/src/github.com/gilbertchen/duplicacy/src/duplicacy_snapshot.go
  222.80MB  2.78% 96.84%   222.80MB  2.78%  github.com/gilbertchen/duplicacy/src.(*SnapshotManager).DownloadSequence /home/s4y/duplicacygo/src/github.com/gilbertchen/duplicacy/src/duplicacy_snapshotmanager.go
  135.01MB  1.68% 98.52%   135.01MB  1.68%  encoding/hex.EncodeToString /usr/lib/go/src/encoding/hex/hex.go
   80.75MB  1.01% 99.53%  7985.55MB 99.55%  github.com/gilbertchen/duplicacy/src.(*SnapshotManager).PruneSnapshots /home/s4y/duplicacygo/src/github.com/gilbertchen/duplicacy/src/duplicacy_snapshotmanager.go
         0     0% 99.53%  8001.56MB 99.75%  github.com/gilbertchen/cli.(*App).Run /home/s4y/duplicacygo/src/github.com/gilbertchen/cli/app.go
         0     0% 99.53%  8001.56MB 99.75%  github.com/gilbertchen/cli.Command.Run /home/s4y/duplicacygo/src/github.com/gilbertchen/cli/command.go
         0     0% 99.53%   135.01MB  1.68%  github.com/gilbertchen/duplicacy/src.(*Config).GetChunkIDFromHash /home/s4y/duplicacygo/src/github.com/gilbertchen/duplicacy/src/duplicacy_config.go
         0     0% 99.53%  7902.80MB 98.52%  github.com/gilbertchen/duplicacy/src.(*SnapshotManager).GetSnapshotChunks /home/s4y/duplicacygo/src/github.com/gilbertchen/duplicacy/src/duplicacy_snapshotmanager.go
         0     0% 99.53%  8001.56MB 99.75%  main.main /home/s4y/duplicacygo/src/github.com/gilbertchen/duplicacy/duplicacy/duplicacy_main.go
         0     0% 99.53%  8001.56MB 99.75%  main.pruneSnapshots /home/s4y/duplicacygo/src/github.com/gilbertchen/duplicacy/duplicacy/duplicacy_main.go
```

**Comment by pdf** [comment]
After this line is where you would need to put `snapshot.ChunkHashes = nil` @s4y :
https://github.com/gilbertchen/duplicacy/blob/8e8a116028ac821a4b3286bb1fa4936d5c013236/src/duplicacy_snapshotmanager.go#L1941

It would obviously be nice to have this fixed properly at some stage though.

**Comment by s4y** [comment]
(@pdf Thanks!) It got further after making that change, but was OOM killed using around 30GB of memory. Before it died, it output a bunch of "Deleting snapshot _x_ at revision _y_" lines and even more "Found unreferenced chunk _hash_" lines.

I'm thinking of trying a "wet run" with `-exclusive -exhaustive` to see if that lets it make progress even if it eventually gets killed, and test if it might be possible to run it several times and get to completion. However, that's kind of a shot in the dark.

**Comment by pdf** [comment]
@s4y as this has continued to be problematic for me on low-resource machines, I've taken a first pass at fixing it in #329.  The changes I've made won't affect exhaustive mode at this stage, but regular prunes should use significantly less memory.

You're welcome to try it out, but use at your own risk until @gilbertchen has time to review, as this is the first time I've touched this code-base, and I have only had a couple of hours free this morning to work on it.
